app/models/commit.rb

```diff
@@ -83,6 +83,18 @@ def refresh_statuses!
     end
   end
 
+  def add_status(status_attributes)
+    commit_was_already_successful = success?
+    new_status = statuses.create!(status_attributes)
+    reload
+    if (!commit_was_already_successful && success?) ||
+       (!new_status.success? && !new_status.pending? &&
+        stack.filter_visible_statuses([new_status]).any? && stack.filter_meaningful_statuses([new_status]).any?)
```

Comment:
I don't think you need such complicated condition. You can query `significant_status` before and after the status creation.


e.g.:

```ruby
previous_status = significant_status
statuses.create!(status_attributes)
new_status = significant_status
if (!previous_status.failure? && new_status.failure?) || (!previous_status.success? && new_status.success?)
  Hook.emit(:commit_status, stack, commit: self, stack: stack, status: new_status)
end
```